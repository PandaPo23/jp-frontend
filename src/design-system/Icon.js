import cleanElement from 'clean-element';
import React from 'react';

import Flex from './Flex';
import system from '../utils/System';
import theme from './theme';

const { icons } = theme;

// Remove `space` props from the `svg` element prevents react warnings
const CleanSvg = cleanElement('svg');

const Base = (props) => {
  const { name, size, ...otherProps } = props;
  const icon = icons[name];
  if (!icon) return <Flex width={size} height={size} />;

  return (
    <Flex centered {...otherProps}>
      {typeof icon === 'function' ? (
        icon(props)
      ) : (
        <CleanSvg
          viewBox={icon.viewBox}
          width={size}
          height={size}
          fill="currentcolor"
        >
          <path d={icon.body[0].props.d} />
        </CleanSvg>
      )}
    </Flex>
  );
};

const Icon = system(
  'Icon',
  {
    extend: Base,
    name: 'check',
    size: 24,
  },
  ({ size, spin }) => ({
    minWidth: size,
    animation: spin && 'spin 2s linear infinite',
  })
);

export default Icon;
