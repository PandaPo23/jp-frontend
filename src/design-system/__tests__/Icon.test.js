import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Icon from '../Icon';

describe('Icon component', () => {
  it('should render icon without crashing', () => {
    const wrapper = mount(<Icon name="menu" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render icon with props', () => {
    const props = {
      name: 'menu',
    };
    const component = mount(<Icon {...props} />);
    expect(component.prop('name')).toBe('menu');
  });
});
