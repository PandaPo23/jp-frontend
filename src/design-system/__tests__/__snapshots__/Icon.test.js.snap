// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component should render icon without crashing 1`] = `
.emotion-0 {
  min-width: 24px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 24px;
}

<Icon
  name="menu"
  size={24}
>
  <Base
    className="emotion-0"
    name="menu"
    size={24}
  >
    <Flex
      centered={true}
      className="emotion-0"
      display="flex"
    >
      <div
        className="emotion-1"
      >
        <svg
          className="emotion-0"
          name="menu"
          size={24}
        />
      </div>
    </Flex>
  </Base>
</Icon>
`;
